src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.cc

```diff
@@ -0,0 +1,569 @@
+#include "onode_node.h"
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+auto node_t<BlockSize, N, NodeType>::key_at(unsigned slot) const
+  -> std::pair<const key_t&, const partial_key_t&>
+{
+  auto& key = keys[slot];
+  if constexpr (item_in_key) {
+    return {key, partial_key_t{}};
+  } else {
+    auto p = from_end(key.offset);
+    return {key, *reinterpret_cast<const partial_key_t*>(p)};
+  }
+}
+
+// update an existing oid with the specified item
+template<size_t BlockSize, int N, ntype_t NodeType>
+ghobject_t
+node_t<BlockSize, N, NodeType>::update_oid_with_slot(unsigned slot,
+                                                     const ghobject_t& oid) const
+{
+  auto [prefix, suffix] = key_at(slot);
+  ghobject_t updated = oid;
+  prefix.update_oid(updated);
+  suffix.update_oid(updated);
+  return updated;
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+auto node_t<BlockSize, N, NodeType>::item_at(const key_t& key) const
+  -> const_item_t
+{
+  if constexpr (item_in_key) {
+    return key.child_addr;
+  } else {
+    assert(key.offset < BlockSize);
+    auto p = from_end(key.offset);
+    auto partial_key = reinterpret_cast<const partial_key_t*>(p);
+    p += size_of(*partial_key);
+    return *reinterpret_cast<const item_t*>(p);
+  }
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+void node_t<BlockSize, N, NodeType>::dump(std::ostream& os) const
+{
+  for (uint16_t i = 0; i < count; i++) {
+    const auto& [prefix, suffix] = key_at(i);
+    os << " [" << i << '/' << count - 1 << "]\n"
+       << "  key1 = (" << prefix << ")\n"
+       << "  key2 = (" << suffix << ")\n";
+    const auto& item = item_at(prefix);
+    if (_is_leaf()) {
+      os << " item = " << item << "\n";
+    } else {
+      os << " child = " << std::hex << item << std::dec << "\n";
+    }
+  }
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+char* node_t<BlockSize, N, NodeType>::from_end(uint16_t offset)
+{
+  auto end = reinterpret_cast<char*>(this) + BlockSize;
+  return end - static_cast<int>(offset);
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+const char* node_t<BlockSize, N, NodeType>::from_end(uint16_t offset) const
+{
+  auto end = reinterpret_cast<const char*>(this) + BlockSize;
+  return end - static_cast<int>(offset);
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+uint16_t node_t<BlockSize, N, NodeType>::used_space() const
+{
+  if constexpr (item_in_key) {
+    return count * sizeof(key_t);
+  } else {
+    if (count) {
+      return keys[count - 1].offset + count * sizeof(key_t);
+    } else {
+      return 0;
+    }
+  }
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+uint16_t node_t<BlockSize, N, NodeType>::capacity()
+{
+  auto p = reinterpret_cast<node_t*>(0);
+  return BlockSize - (reinterpret_cast<char*>(p->keys) -
+                      reinterpret_cast<char*>(p));
```

Comment:
Does offsetof not work here?

Reply:
i had following warning if use `offsetof()`
```
../src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.cc: In instantiation of ‘static uint16_t node_t<BlockSize, N, NodeType>::capacity() [with long unsigned int BlockSize = 512; int N = 3; n\
type_t NodeType = ntype_t::inner; uint16_t = short unsigned int]’:
../src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.cc:566:16:   required from here
../src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.cc:93:31: warning: ‘offsetof’ within non-standard-layout type ‘node_t<512, 3, ntype_t::inner>’ is conditionally-supported [-Winvalid-offs\
etof]
   93 |   return BlockSize - offsetof(node_t, keys);
      |                               ^
```

i think the root cause is that `fixed_key_3` have non-static member variables. but this design simplifies the implementation of `FixedKey<3,...>`. so i think it might be good bargain to trade a homebrew implementation of `offsetof()` with the benefit of a simplified implementation.